File: code/dart/lib/src/core/result.dart
/// Encapsula el resultado de una operación que puede ser exitosa o fallida.
///
/// Usa pattern matching (`switch`) para acceder al valor o al error.
/// El compilador fuerza el manejo exhaustivo de ambos casos:
///
/// ```dart
/// switch (result) {
///   case Success(:final value): // manejar éxito
///   case Failure(:final error): // manejar fallo
/// }
/// ```
sealed class Result<S, F> {
  const Result();

  /// Crea un resultado exitoso con el [value] proporcionado.
  const factory Result.success(S value) = Success<S, F>;

  /// Crea un resultado fallido con el [error] proporcionado.
  const factory Result.failure(F error) = Failure<S, F>;
}

/// Resultado exitoso que contiene el valor de tipo [S].
final class Success<S, F> extends Result<S, F> {
  const Success(this.value);

  /// El valor producido por la operación exitosa.
  final S value;
}

/// Resultado fallido que contiene el error de tipo [F].
final class Failure<S, F> extends Result<S, F> {
  const Failure(this.error);

  /// El error que describe la razón del fallo.
  final F error;
}

File: code/ts/src/core/result.ts
/// Resultado exitoso que contiene el valor de tipo `S`.
export interface Success<S> {
  readonly type: 'success';
  readonly value: S;
}

/// Resultado fallido que contiene el error de tipo `F`.
export interface Failure<F> {
  readonly type: 'failure';
  readonly error: F;
}

/// Encapsula el resultado de una operación que puede ser exitosa o fallida.
///
/// Usa el campo discriminador `type` para pattern matching exhaustivo:
///
/// ```ts
/// switch (result.type) {
///   case 'success': // result.value disponible
///   case 'failure': // result.error disponible
/// }
/// ```
export type Result<S, F> = Success<S> | Failure<F>;

/// Crea un resultado exitoso con el valor proporcionado.
export function success<S, F = never>(value: S): Result<S, F> {
  return { type: 'success', value };
}

/// Crea un resultado fallido con el error proporcionado.
export function failure<S = never, F = unknown>(error: F): Result<S, F> {
  return { type: 'failure', error };
}

/// Type guard que estrecha `Result<S, F>` a `Success<S>`.
export function isSuccess<S, F>(result: Result<S, F>): result is Success<S> {
  return result.type === 'success';
}

/// Type guard que estrecha `Result<S, F>` a `Failure<F>`.
export function isFailure<S, F>(result: Result<S, F>): result is Failure<F> {
  return result.type === 'failure';
}
